File: challenges/notification-dispatcher/python/solution/exercise.py
"""Reference: compose ports without owning their storage or retry policy."""
from tooling.python.contracts import JobQueuePort, KeyValueStore, Sender


class NotificationDispatcher:
    def __init__(self, store: KeyValueStore, queue: JobQueuePort,
                 sender: Sender, idempotency_ttl_ms: int):
        self._store = store
        self._queue = queue
        self._sender = sender
        self._ttl_ms = idempotency_ttl_ms

    def submit(self, key: str, message: str) -> str:
        existing = self._store.get(key)
        if existing is not None:
            return existing
        job_id = self._queue.enqueue(message)
        self._store.put(key, job_id, self._ttl_ms)
        return job_id

    def process_one(self) -> str:
        delivery = self._queue.reserve()
        if delivery is None:
            return "idle"
        try:
            success = self._sender(delivery["payload"])
        except Exception:  # noqa: BLE001 -- Any ordinary sender exception must trigger a retry.
            success = False
        if success:
            self._queue.ack(delivery["receipt"])
            return "sent"
        self._queue.nack(delivery["receipt"])
        return "failed"

File: challenges/notification-dispatcher/go/solution/exercise.go
package exercise

import "refresher/tooling/go/contracts"

type NotificationDispatcher struct {
	store  contracts.KeyValueStore
	queue  contracts.JobQueuePort
	sender contracts.Sender
	ttlMS  int64
}

func NewNotificationDispatcher(store contracts.KeyValueStore, queue contracts.JobQueuePort,
	sender contracts.Sender, idempotencyTTLMS int64) *NotificationDispatcher {
	return &NotificationDispatcher{store, queue, sender, idempotencyTTLMS}
}

func (d *NotificationDispatcher) Submit(key, message string) string {
	if existing, found := d.store.Get(key); found {
		return existing
	}
	id := d.queue.Enqueue(message)
	d.store.Put(key, id, d.ttlMS)
	return id
}

func (d *NotificationDispatcher) ProcessOne() string {
	delivery := d.queue.Reserve()
	if delivery == nil {
		return "idle"
	}
	if err := d.sender(delivery.Payload); err != nil {
		d.queue.Nack(delivery.Receipt)
		return "failed"
	}
	d.queue.Ack(delivery.Receipt)
	return "sent"
}
